### PR TITLE
Workflow variable type PA:CATALOG_OBJECT update validation

### PR DESCRIPTION
**Pull request by tao-xinxiu**
Now PA:CATALOG_OBJECT variable value should exist in the catalog to be validated during workflow submission. (The existence is not checked during workflow design.)

To acheieve this goal `sessionId` is added in ModelValidatorContext (consequently, in the scheduling interface), since it's needed to check catalog object existence.

### CODE REVIEW

File: scheduler/scheduler-api/src/main/java/org/ow2/proactive/scheduler/common/job/factories/spi/model/validator/CatalogObjectValidator.java [hunk 1]
```diff
@@ -56,6 +70,49 @@ public String validate(String parameterValue, ModelValidatorContext context) thr
             throw new ValidationException("Expected value should match regular expression " + pattern.pattern() +
                                           " , received " + parameterValue);
         }
+        if (context == null || context.getSessionId() == null || context.getSessionId().isEmpty()) {
+            // Sometimes the workflow is parsed and checked without scheduler instance (e.g., submitted from catalog).
+            // In this case, we don't have the access of the scheduler global dataspace, so the validity check is passed.
+            logger.debug(String.format("Can't check the validity of the variable value [%s], because missing the access to the catalog",
```
Comment (fviale) [inline scheduler/scheduler-api/src/main/java/org/ow2/proactive/scheduler/common/job/factories/spi/model/validator/CatalogObjectValidator.java:76]: "Validity of the variable value [%s] is skipped because access to the catalog is disabled"

File: scheduler/scheduler-api/src/main/java/org/ow2/proactive/scheduler/common/job/factories/spi/model/validator/CatalogObjectValidator.java [hunk 2]
```diff
@@ -56,6 +70,49 @@ public String validate(String parameterValue, ModelValidatorContext context) thr
...
+                                       parameterValue));
+            return parameterValue;
+        }
+        try {
+            if (!exist(parameterValue, context.getSessionId())) {
+                throw new ValidationException(String.format("Could not find the catalog object [%s].", parameterValue));
```
Comment (fviale) [inline scheduler/scheduler-api/src/main/java/org/ow2/proactive/scheduler/common/job/factories/spi/model/validator/CatalogObjectValidator.java:82]: Catalog object [%s] does not exist or cannot be accessed.

File: scheduler/scheduler-api/src/main/java/org/ow2/proactive/scheduler/common/job/factories/spi/model/validator/CatalogObjectValidator.java [hunk 3]
```diff
@@ -56,6 +70,49 @@ public String validate(String parameterValue, ModelValidatorContext context) thr
...
+            }
+        } catch (PermissionException e) {
+            throw new ValidationException(String.format("Missing the permission to access the catalog object [%s].",
```
Comment (fviale) [inline scheduler/scheduler-api/src/main/java/org/ow2/proactive/scheduler/common/job/factories/spi/model/validator/CatalogObjectValidator.java:85]: Access to catalog object [%s] is not authorized.

File: scheduler/scheduler-api/src/main/java/org/ow2/proactive/scheduler/common/job/factories/spi/model/validator/CatalogObjectValidator.java [hunk 4]
```diff
@@ -56,6 +70,49 @@ public String validate(String parameterValue, ModelValidatorContext context) thr
...
+                                                        parameterValue));
+        } catch (InternalException | IOException e) {
+            logger.warn(String.format("Can't check the validity of the variable value [%s], because failed to request this catalog object.",
```
Comment (fviale) [inline scheduler/scheduler-api/src/main/java/org/ow2/proactive/scheduler/common/job/factories/spi/model/validator/CatalogObjectValidator.java:88]: Cannot check the validity of the variable value [%s]: Internal error.

File: scheduler/scheduler-api/src/main/java/org/ow2/proactive/scheduler/common/job/factories/spi/model/validator/CatalogObjectValidator.java [hunk 5]
```diff
@@ -56,6 +70,49 @@ public String validate(String parameterValue, ModelValidatorContext context) thr
...
+                                      parameterValue),
+                        e);
+            throw new ValidationException(String.format("Failed to request the catalog object [%s].", parameterValue));
```
Comment (fviale) [inline scheduler/scheduler-api/src/main/java/org/ow2/proactive/scheduler/common/job/factories/spi/model/validator/CatalogObjectValidator.java:91]: Internal error when accessing object [%s], please check the server logs.

File: scheduler/scheduler-api/src/main/java/org/ow2/proactive/scheduler/common/job/factories/spi/model/validator/CatalogObjectValidator.java [hunk 6]
```diff
@@ -56,6 +70,49 @@ public String validate(String parameterValue, ModelValidatorContext context) thr
...
+        }
         return parameterValue;
     }
+
+    private boolean exist(String catalogObjectValue, String sessionId) throws PermissionException, IOException {
+        String[] splitCatalog = catalogObjectValue.split("/");
+        String url;
+        // catalogObjectValue is already checked (with the regular expression) to have either 2 to 3 parts (bucketName/objectName[/revision]) after split
+        if (splitCatalog.length == 2) {
```
Comment (fviale) [inline scheduler/scheduler-api/src/main/java/org/ow2/proactive/scheduler/common/job/factories/spi/model/validator/CatalogObjectValidator.java:100]: if the value is not 2 or 3 you should throw a ValidationException

**Review by fviale — APPROVED** [review]
LGTM

File: rest/rest-client/src/test/java/functionaltests/SchedulerClientTest.java [hunk 7]
```diff
@@ -543,7 +543,7 @@ public void testReSubmitJobMergeInfo() throws Exception {
         Map<String, String> info1 = new HashMap<>();
         info1.put("newVar", "newValue");
 
-        JobId jobId1 = client.reSubmit(jobId, Collections.emptyMap(), info1);
+        JobId jobId1 = client.reSubmit(jobId, Collections.emptyMap(), info1, null);
 
         String jobContent1 = client.getJobContent(jobId1);
```
Comment (aminelouati) [inline rest/rest-client/src/test/java/functionaltests/SchedulerClientTest.java:549]: Is it possible to add a functional test with a not null sessionid ? or it is rather a system test to be added ?

**Review by aminelouati — APPROVED** [review]

File: rest/rest-client/src/test/java/functionaltests/SchedulerClientTest.java (hunk 7 above)
Comment (tao-xinxiu) [reply rest/rest-client/src/test/java/functionaltests/SchedulerClientTest.java:549]: Since sessionId is only used in the catalog rest request, the functional test with not-null sessionId will depend on the service of catalog. I think system test would be more proper. I'll add it later. Thank you.
